### PR TITLE
OCPBUGS-34764: Ensure envs are initted upon first load

### PR DESCRIPTION
**Pull request by logonoff**
**Fixes**: 

Maps with JIRA issue https://issues.redhat.com/browse/OCPBUGS-34764

**Analysis / Root cause**: 

There is a data race between the environment variable editor component (which only takes the initial value) and the knative yaml parser. The env variable component loads first which sets the initial value as blank since the parser is still working. Then the parser is done, updates the prop, but is ignored because the env variable component is already loaded. 

Delaying loading the env list for 1 second also fixes the issue...

**Solution Description**: 

Block the env selector from loading until the func.yaml is finished fetching and parsing

**Screen shots / Gifs for design review**: 

n/a

**Unit test coverage report**: 

unchanged

**Test setup:**

Install the Red Hat OpenShift Serverless and setup a Knative service

Then use @Lucifergene 's serverless example https://github.com/Lucifergene/oc-func

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [ ] Safari

### CODE REVIEW

**Comment by Lucifergene** [comment]
@logonoff Please add Unit & E2E tests once completed to prevent the issue resurface again later on :)

**Comment by Lucifergene** [comment]
@logonoff Please rebase this PR.

File: frontend/packages/dev-console/src/components/import/ImportForm.tsx [hunk 1]
```diff
@@ -138,6 +138,7 @@ const ImportForm: React.FC<ImportFormProps & StateProps> = ({
     },
     import: {
       loaded: false,
+      kNativeFuncLoaded: false,
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/ImportForm.tsx:141]: NIT: lets call it `knativeFuncLoaded`. `Knative` is a single word. 
```suggestion
      knativeFuncLoaded: false,
```

File: frontend/packages/dev-console/src/components/import/advanced/DeploymentConfigSection.tsx [hunk 2]
```diff
@@ -22,6 +24,7 @@ const DeploymentConfigSection: React.FC<DeploymentConfigSectionProps> = ({
     values: {
       resources,
       deployment: { env },
+      import: { selectedStrategy, kNativeFuncLoaded: funcLoaded },
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/advanced/DeploymentConfigSection.tsx:27]:
```suggestion
      import: { selectedStrategy, knativeFuncLoaded: funcLoaded },
```

File: frontend/packages/dev-console/src/components/import/advanced/DeploymentConfigSection.tsx [hunk 3]
```diff
@@ -43,12 +46,18 @@ const DeploymentConfigSection: React.FC<DeploymentConfigSectionProps> = ({
...
         />
       )}
-      <EnvironmentField
-        name="deployment.env"
-        label={t('devconsole~Environment variables (runtime only)')}
-        envs={env}
-        obj={deploymentConfigObj}
-      />
+      {(
+        selectedStrategy.type === ImportStrategy.SERVERLESS_FUNCTION ? funcLoaded ?? false : true
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/advanced/DeploymentConfigSection.tsx:50]: Looks nice and compact :+1:

File: frontend/packages/dev-console/src/components/import/import-types.ts [hunk 4]
```diff
@@ -250,6 +250,7 @@ export interface DetectedStrategyFormData extends DetectedStrategy {
 
 export interface ImportStrategyData {
   loaded?: boolean;
+  kNativeFuncLoaded?: boolean;
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/import-types.ts:253]:
```suggestion
  knativeFuncLoaded?: boolean;
```

File: frontend/packages/dev-console/src/components/import/section/build-section/BuildSection.tsx [hunk 5]
```diff
@@ -30,7 +30,7 @@ export const BuildSection: React.FC<BuildSectionProps> = ({ values, appResources
     project: { name: namespace },
     build: { option: buildOption, env: buildEnv },
     image: { selected: selectedImage, tag: selectedTag },
-    import: { selectedStrategy },
+    import: { selectedStrategy, kNativeFuncLoaded: funcLoaded },
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/section/build-section/BuildSection.tsx:33]:
```suggestion
    import: { selectedStrategy, knativeFuncLoaded: funcLoaded },
```

File: frontend/packages/dev-console/src/components/import/section/build-section/__tests__/BuildSection.spec.tsx [hunk 6]
```diff
@@ -102,6 +102,7 @@ describe('BuildSection', () => {
         values={{
           ...componentProps.values,
           import: {
+            kNativeFuncLoaded: true, // env already loaded
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/section/build-section/__tests__/BuildSection.spec.tsx:105]:
```suggestion
            knativeFuncLoaded: true, // env already loaded
```

File: frontend/packages/dev-console/src/components/import/serverless-function/ServerlessFunctionSection.tsx [hunk 7]
```diff
@@ -58,7 +58,10 @@ const ServerlessFunctionSection = ({ builderImages }) => {
           console.warn('Error fetching Serverless Function YAML: ', err);
           setFieldError('ServerlessFunction', err.message);
         })
-        .finally(() => setLoaded(true));
+        .finally(() => {
+          setLoaded(true);
+          setFieldValue('import.kNativeFuncLoaded', true);
```
Comment (Lucifergene) [inline frontend/packages/dev-console/src/components/import/serverless-function/ServerlessFunctionSection.tsx:63]:
```suggestion
          setFieldValue('import.knativeFuncLoaded', true);
```

File: frontend/packages/console-shared/src/components/formik-fields/EnvironmentField.tsx [hunk 8]
```diff
@@ -26,7 +26,7 @@ const EnvironmentField: React.FC<EnvironmentFieldProps> = ({
   const { t } = useTranslation();
   const fieldId = getFieldId(props.name, 'env-input');
   const environmentVariables = React.useMemo(() => {
-    return !_.isEmpty(envs) ? envs.map((env) => _.values(env)) : [['', '']];
+    return _.isEmpty(envs) ? [['', '']] : envs.map((env) => _.values(env));
```
Comment (Lucifergene) [inline frontend/packages/console-shared/src/components/formik-fields/EnvironmentField.tsx:29]: Nothing bad, just it increases the code diff. 
Should be avoided if not very necessary.
But we can move forward with ur change this time :+1:

File: frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature
Comment (Lucifergene) [inline frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature:1]: Thanks for updating the E2E test for this.
Hoping this does not break anymore in the future :)

**Review by Lucifergene — CHANGES_REQUESTED** [review]
Thanks @logonoff for fixing this bug. It was pending for a long time.
I have added some small NITs. PTAL.

File: frontend/packages/console-shared/src/components/formik-fields/EnvironmentField.tsx (hunk 8 above)
Comment (logonoff) [reply frontend/packages/console-shared/src/components/formik-fields/EnvironmentField.tsx:29]: in my defense it took me a few minutes to parse out the original code and I wasn't going to open a PR to refactor one line :p

**Review by vikram-raj — APPROVED** [review]
/approve

**Comment by sanketpathak** [comment]
@logonoff The updated test is failing 
Screenshot 2024-08-27 at 3 31 50 AM

File: frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature [hunk 9]
```diff
@@ -1,23 +1,26 @@
...
+        Scenario: Create knative workload from Import From Git card on Add page: EE-02-TC01
+            Given user is at Add page
+              And user is at Import from Git form
+             When user enters Git Repo URL as "https://github.com/logonoff/oc-node-func-with-env"
+              And user enters Name as "node-knative"
+              And user clicks "Show advanced Build option" link in Advanced Options section
+              And user clicks "Show advanced Deployment option" link in Advanced Options section
+             Then the environment variable "MY_BUILD_KEY" has value "tests" in the advanced options of the Build section in the Import from Git page
+             Then the environment variable "MY_API_KEY" has value "{{ env:API_KEY }}" in the advanced options of the Deployment section in the Import from Git page
+              And user clicks Create button on Add page
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature:21]: In the gherkin script, we do not use the `Then` keyword twice, use `And` to replace it. We do not perform any actions in the `Then` statement, we only check the results. As we already have a working knative creation scenario this test case can be rewritten as:
```suggestion
        Scenario: Check environment variable initialisation in git form : EE-02-TC01
            Given user is at Add page
              And user is at Import from Git form
             When user enters Git Repo URL as "https://github.com/logonoff/oc-node-func-with-env"
              And user enters Name as "node-knative"
              And user clicks "Show advanced Build option" link in Advanced Options section
              And user clicks "Show advanced Deployment option" link in Advanced Options section
             Then the environment variable "MY_BUILD_KEY" has value "tests" in the advanced options of the Build section in the Import from Git page
              And the environment variable "MY_API_KEY" has value "{{ env:API_KEY }}" in the advanced options of the Deployment section in the Import from Git page
```

File: frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature [hunk 10]
```diff
@@ -1,23 +1,26 @@
...
              Then user will be redirected to Topology page
-              And user is able to see workload "knative-demo" in topology page list view
+              And user is able to see workload "node-knative" in topology page list view
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature:23]: These lines can be removed

File: frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature (hunk 9 above)
Comment (logonoff) [reply frontend/packages/dev-console/integration-tests/features/e2e/knative-service-workload.feature:21]: thanks for the fix :p

**Comment by sanketpathak** [comment]
/lgtm
/label qe-approved
Tests work fine
Screenshot 2024-08-28 at 7 48 45 PM

**Pull request merged**